packages/editor/src/components/provider/content-only-lock-sections.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch, useRegistry } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { useEffect } from '@wordpress/element';
+import { store as blocksStore } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+
+/**
+ * Component that when rendered, makes it so that the editor allows only
+ * specific sections to be edited in content-only mode for templates.
+ */
+export default function ContentOnlyLockSections() {
+	const registry = useRegistry();
+
+	const { setBlockEditingMode, unsetBlockEditingMode } =
+		useDispatch( blockEditorStore );
+
+	const selected = useSelect( ( select ) => {
+		const {
+			getClientIdsOfDescendants,
+			getClientIdsWithDescendants,
+			getSectionRootClientId,
+			getSectionClientIds,
+		} = unlock( select( blockEditorStore ) );
+
+		const sectionRootClientId = getSectionRootClientId();
+		const sectionClientIds = getSectionClientIds();
+		const allClientIds = sectionRootClientId
+			? getClientIdsOfDescendants( sectionRootClientId )
+			: getClientIdsWithDescendants();
+		return {
+			sectionRootClientId,
+			sectionClientIds,
+			allClientIds,
+		};
+	}, [] );
+	const { sectionClientIds, allClientIds, sectionRootClientId } = selected;
+	const { getBlockOrder, getBlockName } = useSelect( blockEditorStore );
+	const { __experimentalHasContentRoleAttribute } = useSelect( blocksStore );
+
+	useEffect( () => {
+		const sectionChildrenClientIds = sectionClientIds.flatMap(
+			( clientId ) => getBlockOrder( clientId )
+		);
+		const contentClientIds = allClientIds.filter( ( clientId ) =>
+			__experimentalHasContentRoleAttribute( getBlockName( clientId ) )
+		);
+
+		registry.batch( () => {
+			// 1. The section root should hide non-content controls.
+			setBlockEditingMode( sectionRootClientId, 'contentOnly' );
+
+			// 2. Each section should hide non-content controls.
+			for ( const clientId of sectionClientIds ) {
+				setBlockEditingMode( clientId, 'contentOnly' );
+			}
+
+			// 3. The children of the section should be disabled.
+			for ( const clientId of sectionChildrenClientIds ) {
+				setBlockEditingMode( clientId, 'disabled' );
+			}
+
+			// 4. ...except for the "content" blocks which should be content-only.
+			for ( const clientId of contentClientIds ) {
+				setBlockEditingMode( clientId, 'contentOnly' );
+			}
```

Comment:
Can we start moving away from `setBlockEditingMode` usage and instead update the "getBlockEditingMode" selector to guess the right mode based on the state instead?

Reply:
I see. So rather than batching multiple actions, instead we make the selector more complex so that it can infer the correct block editing mode on a per block basis?

Reply:
yes, that makes the code more solid IMO. because it's easy for folks to render competing components like the `ContentOnlyLockSections` here and you end up with race conditions. For instance, I know we had the navigation block at some point call `setBlockEditingMode` directly which conflicted with the locking applied based on the "editor package rendering mode" and the more we add this kind of components, the worse it gets.

Reply:
I'll take a look at this. I'd also love to solve the similar situation in the pattern block, though I wonder if it'd be ok to build so much into one selector.

Reply:
@talldan worth noting that I already implemented this for the "select mode" in this PR https://github.com/WordPress/gutenberg/pull/65204

Reply:
The idea was that folks would use `useBlockEditingMode` in their blocks but yeah unfortunately it's become more common to call `setBlockEditingMode` directly which has nasty side effects. Need to revisit this API in favour of something more declarative. Maybe an array of "locks" that is passed in via block settings.